File: Unity_scripts/ObjectPositionModifier.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using uPLibrary.Networking.M2Mqtt;
using uPLibrary.Networking.M2Mqtt.Messages;
using System.Text;
using Newtonsoft.Json;

public class ObjectPositionModifier : MonoBehaviour
{
    private MqttClient client;
    private string brokerHostname = ""; // Add the broker host IP here
    private int brokerPort = 0000; // Add the broker port here
    public string objectID;
    public string roomID;

    private string publishTopic;
    private string subscribeTopic;

    // Store the last published position to compare with the current position
    private Vector3 lastPublishedPosition;
    private float lastPublishedRotation;

    // private Queue<Action> mainThreadActions = new Queue<Action>();


    // Start is called before the first frame update
    void Start()
    {
        client = new MqttClient(brokerHostname, brokerPort, false, null, null, MqttSslProtocols.None);
        client.MqttMsgPublishReceived += OnMessageReceived;
        string clientId = System.Guid.NewGuid().ToString();
        client.Connect(clientId);

        // Define topics
        publishTopic = $"{roomID}/{objectID}/transform";
        subscribeTopic = $"config/{roomID}/{objectID}/transform";

        client.Subscribe(new string[] { subscribeTopic }, new byte[] { MqttMsgBase.QOS_LEVEL_AT_LEAST_ONCE });

        // Publish object's position immediately when the script starts
        PublishTransform(transform.position, transform.rotation.eulerAngles.z);

        lastPublishedPosition = transform.position;
        lastPublishedRotation = NormalizeAngle(transform.rotation.eulerAngles.z);

        StartCoroutine(CheckAndPublishTransform());
    }

    void OnMessageReceived(object sender, MqttMsgPublishEventArgs e)
    {
        string message = Encoding.UTF8.GetString(e.Message);

        string[] transformData = message.Split(',');

        if (transformData.Length == 4)
        {
            // Parse position data (x, y, z)
            float x = float.Parse(transformData[0]);
            float y = float.Parse(transformData[1]);
            float z = float.Parse(transformData[2]);
            // Parse rotation data (z)
            float rotZ = float.Parse(transformData[3]);

            // Convert the angle from degrees to radians
            float rotZRadians = rotZ * Mathf.Deg2Rad;

            // Calculate the quaternion for the rotation around the Z-axis
            float cos = Mathf.Cos(rotZRadians / 2);
            float sin = Mathf.Sin(rotZRadians / 2);
            Quaternion rotationQuat = new Quaternion(0, 0, sin, cos);

            // // Update the object's position
            // transform.position = new Vector3(x, y, z);

            // Run the position update on the main thread using UnityMainThreadDispatcher
            UnityMainThreadDispatcher.Instance().Enqueue(() => {
                transform.position = new Vector3(x, y, z);
                transform.rotation = rotationQuat;
            });
        }
    }

    IEnumerator CheckAndPublishTransform() {

        while (true) {
            // Get current position and rotation of the object
            Vector3 currentPosition = transform.position;
            float currentRotation = NormalizeAngle(transform.rotation.eulerAngles.z);

            // Check if position is changed
            if (!currentPosition.Equals(lastPublishedPosition) || !currentRotation.Equals(lastPublishedRotation)) {
                
                // // Create the message in format "x,y,z"
                // string message = $"{currentPosition.x},{currentPosition.y},{currentPosition.z}";

                // // Publish the message
                // client.Publish(publishTopic, Encoding.UTF8.GetBytes(message), MqttMsgBase.QOS_LEVEL_AT_LEAST_ONCE, false);

                // Publish the new position if it has changed
                PublishTransform(currentPosition, currentRotation);
                
                // Update the last published position
                lastPublishedPosition = currentPosition;
                lastPublishedRotation = currentRotation;
            }

            // wait for 0.5 seconds before checking again
            yield return new WaitForSeconds(0.5f);
        }
    }

    float NormalizeAngle(float angle)
    {
        if (angle > 180)
            angle -= 360;
        return angle;
    }

    void PublishTransform(Vector3 position, float rotation) {
        
        // Create the message in format "x,y,z"
        string message = $"{position.x},{position.y},{position.z},{rotation}";

        // Publish the message
        client.Publish(publishTopic, Encoding.UTF8.GetBytes(message), MqttMsgBase.QOS_LEVEL_AT_LEAST_ONCE, false);
    }

    void OnDestroy() {
        // Disconnect the client when the object is destroyed
        if (client.IsConnected) {
            client.Disconnect();
        }
    }
}

File: Unity_scripts/Object_position_modifier.py
import paho.mqtt.client as mqtt
import time

# MQTT broker settings
broker_address = ''  # Replace with your MQTT broker address
port = 1883  # Typically 1883 for unencrypted MQTT communication

# The topic to which the Unity object is subscribed
objectID = "Couch"  # Replace with the object ID from Unity
roomID = "Hallway"  # Replace with the room ID from Unity
publish_topic = f"config/{roomID}/{objectID}/transform"  # Topic format used in Unity

# Create a new MQTT client instance
client = mqtt.Client()

# Connect to the MQTT broker
client.connect(broker_address, port)

# Function to publish new position to the topic
def publish_new_position(x, y, z, rotZ):
    # Format the position as "x,y,z"
    message = f"{x},{y},{z},{rotZ}"
    
    # Publish the message to the topic
    client.publish(publish_topic, message)
    print(f"Published new position to {publish_topic}: {message}")

# Example of publishing new positions
def main():
    try:
        while True:
            # Example: update new position (you can modify this with real data)
            x, y, z, rotZ = 2.07, -7.056, -1, 0  # Example new position
            
            # Publish the new position
            publish_new_position(x, y, z, rotZ)

            # Wait for 5 seconds before sending the next position
            time.sleep(5)

    except KeyboardInterrupt:
        print("Stopping script")
    
    finally:
        # Disconnect MQTT client
        client.disconnect()
        client.loop_stop()

if __name__ == "__main__":
    main()



# import time
# import json
# import paho.mqtt.client as mqtt
# import numpy as np

# broker_address = ''
# broker_port = 1883

# def on_connect(client, userdata, flags, rc):
#     print(f'Connected with result code {str(rc)}')
#     client.subscribe("activity/Finished")

# def on_message(client, userdata, msg):
#     global activity_finished
#     try:
#         payload = msg.payload.decode()
#         if payload == 'True':
#             activity_finished = True
#         elif payload == 'False':
#             activity_finished = False
#         print(f'Received activityFinished: {activity_finished}')
#     except Exception as e:
#         print('Error decoding JSON: {e}')

# # Initialize MQTT client
# client = mqtt.Client()
# client.on_connect = on_connect
# client.on_message = on_message

# # connect to MQTT broker
# client.connect(broker_address, broker_port, 60)

# client.loop_start()

# try:
#     while True:
#         # Get the current position and activity time
#         current_data = data[i]
#         current_position = {"x": current_data[0], "y": current_data[1], "z": current_data[2]}
#         current_activity_time = current_data[3]

#         if activity_finished:

#             # Publish position to target topic
#             position_topic = "activity/position"
#             client.publish(position_topic, json.dumps(current_position))
#             # Publish activity time to unit topic
#             activity_time_topic = "activity/time"
#             client.publish(activity_time_topic, str(current_activity_time))

#             i = (i + 1) % len(data)
#             # print(i)

#         time.sleep(1)

# except KeyboardInterrupt:
#     print('Interrupted, exiting...')
# finally:
#     # Disconnect MQTT client
#     client.disconnect()
#     client.loop_stop()